### PR TITLE
remove peerDependency on graphql 14

### PR DESCRIPTION
**Pull request by phryneas**
This removes `graphql@14` from our peerDependencies (see #9167) and also adds a CI run that installs different versions of our peerDependencies and runs `tsc` to see if we have any type conflicts with any of them. (At first also including `graphql` 14 as that should definitely fail.)

### Checklist:

- [ ] If this PR contains changes to the library itself (not necessary for e.g. docs updates), please include a changeset (see [CONTRIBUTING.md](https://github.com/apollographql/apollo-client/blob/main/CONTRIBUTING.md#changesets))
- [ ] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [ ] Make sure all of the significant new logic is covered by tests

### CODE REVIEW

**Comment by phryneas** [comment]
So, from that: our internal types are definitely not compatible to all of those. I'll see that I can set up that test differently to actually test our shipped types.

**Comment by phryneas** [comment]
Yup, that works as expected:
image
[Link to Workflow with `graphql@14` failing](https://app.circleci.com/pipelines/github/apollographql/apollo-client/21487/workflows/eb4ff3df-7efe-4f30-b3f2-a234b1dc3b37)

=> I'll remove `graphql@14` from the matrix

File: .circleci/config.yml [hunk 1]
```diff
@@ -109,3 +127,17 @@ workflows:
...
+      - TestPeerDepTypes:
+          name: Test external types for << matrix.externalPackage >>
+          requires:
+            - BuildTarball
+          matrix:
+            parameters:
+              externalPackage:
+                - "graphql@15"
+                - "graphql@16"
+                - "graphql@^17.0.0-alpha"
```
Comment (phryneas) [inline .circleci/config.yml:139]: might be good to have an eye on this

File: .circleci/config.yml [hunk 2]
```diff
@@ -109,3 +127,17 @@ workflows:
...
+                - "@types/react@16.8 @types/react-dom@16.8"
+                - "@types/react@17 @types/react-dom@17"
+                - "@types/react@18 @types/react-dom@18"
+                - "typescript@next"
```
Comment (phryneas) [inline .circleci/config.yml:143]: and this

**Review by alessbell — APPROVED** [review]
+1 to a lightweight test matrix here for added visibility 🚀

**Pull request merged**